File: src/main/java/ltdjms/discord/shop/services/EcpayCvsPaymentService.java
package ltdjms.discord.shop.services;

import java.net.URI;
import java.net.URLDecoder;
import java.net.URLEncoder;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.security.GeneralSecurityException;
import java.time.Clock;
import java.time.Duration;
import java.time.Instant;
import java.time.LocalDateTime;
import java.time.ZoneId;
import java.time.format.DateTimeFormatter;
import java.util.Base64;
import java.util.Locale;
import java.util.Objects;
import javax.crypto.Cipher;
import javax.crypto.spec.IvParameterSpec;
import javax.crypto.spec.SecretKeySpec;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ObjectNode;

import ltdjms.discord.shared.DomainError;
import ltdjms.discord.shared.EnvironmentConfig;
import ltdjms.discord.shared.Result;

/** Calls ECPay backend API to generate CVS payment code for fiat orders. */
public class EcpayCvsPaymentService {

  private static final Logger LOG = LoggerFactory.getLogger(EcpayCvsPaymentService.class);

  private static final String STAGE_ENDPOINT =
      "https://ecpayment-stage.ecpay.com.tw/1.0.0/Cashier/GenPaymentCode";
  private static final String PROD_ENDPOINT =
      "https://ecpayment.ecpay.com.tw/1.0.0/Cashier/GenPaymentCode";
  private static final DateTimeFormatter TRADE_DATE_FORMAT =
      DateTimeFormatter.ofPattern("yyyy/MM/dd HH:mm:ss");
  private static final DateTimeFormatter MERCHANT_TRADE_NO_TIME_FORMAT =
      DateTimeFormatter.ofPattern("yyMMddHHmmssSSS");

  private final EnvironmentConfig config;
  private final HttpClient httpClient;
  private final ObjectMapper objectMapper;
  private final Clock clock;
  private long lastTradeNoMillis = -1L;
  private int tradeNoSequence = 0;

  public EcpayCvsPaymentService(EnvironmentConfig config) {
    this(config, HttpClient.newHttpClient(), new ObjectMapper(), Clock.systemDefaultZone());
  }

  EcpayCvsPaymentService(
      EnvironmentConfig config, HttpClient httpClient, ObjectMapper objectMapper, Clock clock) {
    this.config = Objects.requireNonNull(config, "config must not be null");
    this.httpClient = Objects.requireNonNull(httpClient, "httpClient must not be null");
    this.objectMapper = Objects.requireNonNull(objectMapper, "objectMapper must not be null");
    this.clock = Objects.requireNonNull(clock, "clock must not be null");
  }

  /**
   * Generates a CVS payment code from ECPay.
   *
   * @param totalAmountTwd payment amount in TWD
   * @param itemName order item name
   * @param tradeDesc order description
   * @return generated payment code data or error
   */
  public Result<CvsPaymentCode, DomainError> generateCvsPaymentCode(
      long totalAmountTwd, String itemName, String tradeDesc) {
    if (totalAmountTwd <= 0) {
      return Result.err(DomainError.invalidInput("法幣付款金額必須大於 0"));
    }
    if (itemName == null || itemName.isBlank()) {
      return Result.err(DomainError.invalidInput("商品名稱不能為空"));
    }

    String merchantId = config.getEcpayMerchantId();
    String hashKey = config.getEcpayHashKey();
    String hashIv = config.getEcpayHashIv();
    String returnUrl = config.getEcpayReturnUrl();
    if (isBlank(merchantId) || isBlank(hashKey) || isBlank(hashIv) || isBlank(returnUrl)) {
      return Result.err(
          DomainError.invalidInput("綠界金流尚未完成設定（MerchantID/HashKey/HashIV/ReturnURL）"));
    }

    String merchantTradeNo = generateMerchantTradeNo();

    try {
      String dataPayload =
          buildRequestDataPayload(
              merchantId,
              merchantTradeNo,
              totalAmountTwd,
              itemName.trim(),
              tradeDesc == null || tradeDesc.isBlank() ? "Discord 商品下單" : tradeDesc,
              returnUrl,
              clampCvsExpireMinutes(config.getEcpayCvsExpireMinutes()));
      String encryptedData = encryptData(dataPayload, hashKey, hashIv);

      ObjectNode root = objectMapper.createObjectNode();
      root.put("MerchantID", merchantId);
      root.putObject("RqHeader").put("Timestamp", Instant.now(clock).getEpochSecond());
      root.put("Data", encryptedData);

      HttpRequest request =
          HttpRequest.newBuilder()
              .uri(URI.create(config.getEcpayStageMode() ? STAGE_ENDPOINT : PROD_ENDPOINT))
              .timeout(Duration.ofSeconds(15))
              .header("Content-Type", "application/json")
              .POST(HttpRequest.BodyPublishers.ofString(objectMapper.writeValueAsString(root)))
              .build();

      HttpResponse<String> response =
          httpClient.send(request, HttpResponse.BodyHandlers.ofString(StandardCharsets.UTF_8));

      if (response.statusCode() != 200) {
        LOG.warn(
            "ECPay request failed with status={} body={}", response.statusCode(), response.body());
        return Result.err(DomainError.unexpectedFailure("綠界服務暫時不可用，請稍後再試", null));
      }

      JsonNode responseNode = objectMapper.readTree(response.body());
      int transCode = responseNode.path("TransCode").asInt(-1);
      if (transCode != 1) {
        String transMsg = responseNode.path("TransMsg").asText("未知錯誤");
        LOG.warn("ECPay transCode failed: transCode={} transMsg={}", transCode, transMsg);
        return Result.err(DomainError.unexpectedFailure("綠界取號失敗：" + transMsg, null));
      }

      String encryptedResponseData = responseNode.path("Data").asText("");
      if (encryptedResponseData.isBlank()) {
        LOG.warn("ECPay response data is empty: {}", response.body());
        return Result.err(DomainError.unexpectedFailure("綠界回傳資料不完整", null));
      }

      String decryptedJson = decryptData(encryptedResponseData, hashKey, hashIv);
      JsonNode dataNode = objectMapper.readTree(decryptedJson);

      int rtnCode = dataNode.path("RtnCode").asInt(-1);
      if (rtnCode != 1) {
        String rtnMsg = dataNode.path("RtnMsg").asText("未知錯誤");
        LOG.warn("ECPay business failed: rtnCode={} rtnMsg={}", rtnCode, rtnMsg);
        return Result.err(DomainError.unexpectedFailure("綠界取號失敗：" + rtnMsg, null));
      }

      JsonNode orderInfo = dataNode.path("OrderInfo");
      String orderNumber = orderInfo.path("MerchantTradeNo").asText("");
      JsonNode cvsInfo = dataNode.path("CVSInfo");
      String paymentNo = cvsInfo.path("PaymentNo").asText("");
      String expireDate = textOrNull(cvsInfo.path("ExpireDate").asText(null));
      String paymentUrl = textOrNull(cvsInfo.path("PaymentURL").asText(null));

      if (orderNumber.isBlank() || paymentNo.isBlank()) {
        LOG.warn("ECPay response missing orderNumber or paymentNo: {}", decryptedJson);
        return Result.err(DomainError.unexpectedFailure("綠界回傳資料不完整", null));
      }

      return Result.ok(new CvsPaymentCode(orderNumber, paymentNo, expireDate, paymentUrl));
    } catch (InterruptedException e) {
      Thread.currentThread().interrupt();
      LOG.warn("ECPay request interrupted", e);
      return Result.err(DomainError.unexpectedFailure("綠界連線被中斷，請稍後再試", e));
    } catch (Exception e) {
      LOG.error("Failed to generate ECPay CVS payment code", e);
      return Result.err(DomainError.unexpectedFailure("建立法幣訂單失敗，請稍後再試", e));
    }
  }

  private String buildRequestDataPayload(
      String merchantId,
      String merchantTradeNo,
      long totalAmountTwd,
      String itemName,
      String tradeDesc,
      String returnUrl,
      int cvsExpireMinutes)
      throws Exception {
    ObjectNode data = objectMapper.createObjectNode();
    data.put("MerchantID", merchantId);
    data.put("ChoosePayment", "CVS");

    ObjectNode orderInfo = data.putObject("OrderInfo");
    orderInfo.put(
        "MerchantTradeDate",
        TRADE_DATE_FORMAT.format(LocalDateTime.now(clock.withZone(ZoneId.systemDefault()))));
    orderInfo.put("MerchantTradeNo", merchantTradeNo);
    orderInfo.put("TotalAmount", totalAmountTwd);
    orderInfo.put("ReturnURL", returnUrl);
    orderInfo.put("TradeDesc", tradeDesc);
    orderInfo.put("ItemName", itemName);

    ObjectNode cvsInfo = data.putObject("CVSInfo");
    cvsInfo.put("ExpireDate", cvsExpireMinutes);
    cvsInfo.put("CVSCode", "CVS");

    return objectMapper.writeValueAsString(data);
  }

  private String encryptData(String plainJson, String hashKey, String hashIv)
      throws GeneralSecurityException {
    String urlEncoded = URLEncoder.encode(plainJson, StandardCharsets.UTF_8);
    // ECPay requires fixed merchant HashKey/HashIV with AES/CBC for request encryption.
    Cipher cipher = Cipher.getInstance("AES/CBC/PKCS5Padding");
    SecretKeySpec keySpec = new SecretKeySpec(hashKey.getBytes(StandardCharsets.UTF_8), "AES");
    IvParameterSpec ivSpec = new IvParameterSpec(hashIv.getBytes(StandardCharsets.UTF_8));
    cipher.init(Cipher.ENCRYPT_MODE, keySpec, ivSpec);
    byte[] encryptedBytes = cipher.doFinal(urlEncoded.getBytes(StandardCharsets.UTF_8));
    return Base64.getEncoder().encodeToString(encryptedBytes);
  }

  private String decryptData(String encryptedData, String hashKey, String hashIv)
      throws GeneralSecurityException {
    // ECPay response decryption uses the same fixed merchant HashKey/HashIV rule.
    Cipher cipher = Cipher.getInstance("AES/CBC/PKCS5Padding");
    SecretKeySpec keySpec = new SecretKeySpec(hashKey.getBytes(StandardCharsets.UTF_8), "AES");
    IvParameterSpec ivSpec = new IvParameterSpec(hashIv.getBytes(StandardCharsets.UTF_8));
    cipher.init(Cipher.DECRYPT_MODE, keySpec, ivSpec);
    byte[] decoded = Base64.getDecoder().decode(encryptedData);
    byte[] decrypted = cipher.doFinal(decoded);
    String urlEncodedPlain = new String(decrypted, StandardCharsets.UTF_8);
    return URLDecoder.decode(urlEncodedPlain, StandardCharsets.UTF_8);
  }

  private int clampCvsExpireMinutes(int input) {
    if (input < 1) {
      return 1;
    }
    return Math.min(input, 43200);
  }

  private synchronized String generateMerchantTradeNo() {
    long currentMillis = Instant.now(clock).toEpochMilli();
    if (currentMillis < lastTradeNoMillis) {
      currentMillis = lastTradeNoMillis;
    }

    if (currentMillis == lastTradeNoMillis) {
      tradeNoSequence++;
      if (tradeNoSequence > 999) {
        currentMillis = lastTradeNoMillis + 1;
        tradeNoSequence = 0;
      }
    } else {
      tradeNoSequence = 0;
    }

    lastTradeNoMillis = currentMillis;
    String timePart =
        MERCHANT_TRADE_NO_TIME_FORMAT.format(
            LocalDateTime.ofInstant(Instant.ofEpochMilli(currentMillis), ZoneId.systemDefault()));
    String sequencePart = String.format(Locale.ROOT, "%03d", tradeNoSequence);
    return "FD" + timePart + sequencePart;
  }

  private boolean isBlank(String value) {
    return value == null || value.isBlank();
  }

  private String textOrNull(String value) {
    return value == null || value.isBlank() ? null : value;
  }

  /** CVS payment code info returned from ECPay. */
  public record CvsPaymentCode(
      String orderNumber, String paymentNo, String expireDate, String paymentUrl) {}
}

File: packages/shop/src/services/ecpay-cvs-payment.service.ts
import { encryptAES, decryptAES } from '../crypto/ecpay-aes.js';
import type { EnvironmentConfig } from '@ltdjms/shared';
import { Result, ok, err, DomainError } from '@ltdjms/shared';
import { Agent } from 'undici';
import { fetchWithRetry } from './fetch-retry.js';
import crypto from 'node:crypto';
import pino from 'pino';

const STAGE_ENDPOINT = 'https://ecpayment-stage.ecpay.com.tw/1.0.0/Cashier/GenPaymentCode';
const PROD_ENDPOINT = 'https://ecpayment.ecpay.com.tw/1.0.0/Cashier/GenPaymentCode';
// ECPay official stage (test) credentials — these are PUBLIC test credentials.
// They are embedded here for stage-mode detection: the application compares
// the user's .env values against these constants and refuses to start in
// production mode (ECPAY_STAGE_MODE=false) if stage credentials are detected.
// DO NOT add production credentials here — they belong in .env only.
const OFFICIAL_STAGE_MERCHANT_ID = '3002607';
const OFFICIAL_STAGE_HASH_KEY = 'pwFHCqoQZGmho4w6';
const OFFICIAL_STAGE_HASH_IV = 'EkRm7iFT261dpevs';

const keepAliveDispatcher = new Agent({ keepAliveTimeout: 10 });

function pad2(n: number): string {
  return n.toString().padStart(2, '0');
}

function pad3(n: number): string {
  return n.toString().padStart(3, '0');
}

export interface CvsPaymentCode {
  orderNumber: string;
  paymentNo: string;
  expireDate: string | null;
  expireAt: Date;
  paymentUrl: string | null;
}

export class EcpayCvsPaymentService {
  private readonly log: pino.Logger;

  constructor(
    private readonly config: EnvironmentConfig,
    logger?: pino.Logger,
  ) {
    this.log = logger ?? pino({ level: 'warn' });
  }

  async generateCvsPaymentCode(
    totalAmountTwd: number,
    itemName: string,
    tradeDesc: string,
  ): Promise<Result<CvsPaymentCode, DomainError>> {
    if (totalAmountTwd <= 0) {
      return err(DomainError.invalidInput('法幣付款金額必須大於 0'));
    }
    if (!itemName || itemName.trim().length === 0) {
      return err(DomainError.invalidInput('商品名稱不能為空'));
    }

    const merchantId = this.config.getEcpayMerchantId().trim();
    const hashKey = this.config.getEcpayHashKey().trim();
    const hashIv = this.config.getEcpayHashIv().trim();
    const returnUrl = this.config.getEcpayReturnUrl().trim();

    if (!merchantId || !hashKey || !hashIv || !returnUrl) {
      return err(
        DomainError.invalidInput('綠界金流尚未完成設定（MerchantID/HashKey/HashIV/ReturnURL）'),
      );
    }

    if (
      merchantId === OFFICIAL_STAGE_MERCHANT_ID &&
      hashKey === OFFICIAL_STAGE_HASH_KEY &&
      hashIv === OFFICIAL_STAGE_HASH_IV &&
      !this.config.getEcpayStageMode()
    ) {
      return err(
        DomainError.invalidInput(
          '目前設定的是綠界官方測試 MerchantID/HashKey/HashIV，但 ECPAY_STAGE_MODE=false。請切回測試環境或改用正式環境金鑰。',
        ),
      );
    }

    const requestAt = new Date();
    const merchantTradeNo = this.generateMerchantTradeNo();

    try {
      const dataPayload = this.buildRequestDataPayload(
        merchantId,
        merchantTradeNo,
        totalAmountTwd,
        itemName.trim(),
        tradeDesc && tradeDesc.trim().length > 0 ? tradeDesc.trim() : 'Discord 商品下單',
        returnUrl,
        this.clampCvsExpireMinutes(this.config.getEcpayCvsExpireMinutes()),
      );

      const encryptedData = encryptAES(dataPayload, hashKey, hashIv);

      const root = {
        MerchantID: merchantId,
        RqHeader: { Timestamp: Math.floor(Date.now() / 1000) },
        Data: encryptedData,
      };

      const endpoint = this.config.getEcpayStageMode() ? STAGE_ENDPOINT : PROD_ENDPOINT;

      const response = await fetchWithRetry(endpoint, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(root),
        signal: AbortSignal.timeout(15000),
        dispatcher: keepAliveDispatcher,
      });

      if (!response.ok) {
        const body = await response.text();
        this.log.warn({ status: response.status, body }, 'ECPay request failed');
        return err(DomainError.unexpectedFailure('綠界服務暫時不可用，請稍後再試'));
      }

      const responseJson: any = await response.json();
      const transCode = responseJson.TransCode ?? -1;
      if (transCode !== 1) {
        const transMsg = responseJson.TransMsg ?? '未知錯誤';
        this.log.warn(
          { transCode, transMsg, merchantId, stageMode: this.config.getEcpayStageMode() },
          'ECPay transCode failed',
        );
        return err(DomainError.unexpectedFailure(this.buildTransCodeFailureMessage(transMsg)));
      }

      const encryptedResponseData: string = responseJson.Data ?? '';
      if (!encryptedResponseData) {
        this.log.warn({ body: responseJson }, 'ECPay response data is empty');
        return err(DomainError.unexpectedFailure('綠界回傳資料不完整'));
      }

      let decryptedJson: string;
      try {
        decryptedJson = decryptAES(encryptedResponseData, hashKey, hashIv);
      } catch (e) {
        this.log.warn({ error: e }, 'Failed to decrypt ECPay response');
        return err(DomainError.unexpectedFailure('綠界回傳資料解密失敗'));
      }

      const dataNode: any = JSON.parse(decryptedJson);
      const rtnCode = dataNode.RtnCode ?? -1;
      if (rtnCode !== 1) {
        const rtnMsg = dataNode.RtnMsg ?? '未知錯誤';
        this.log.warn({ rtnCode, rtnMsg }, 'ECPay business failed');
        return err(DomainError.unexpectedFailure(`綠界取號失敗：${rtnMsg}`));
      }

      const orderInfo = dataNode.OrderInfo ?? {};
      const orderNumber: string = orderInfo.MerchantTradeNo ?? '';
      const cvsInfo = dataNode.CVSInfo ?? {};
      const paymentNoVal: string = cvsInfo.PaymentNo ?? '';
      const expireDateStr: string | null = cvsInfo.ExpireDate ?? null;
      const expireAt = this.resolveExpireAt(
        expireDateStr,
        requestAt,
        this.clampCvsExpireMinutes(this.config.getEcpayCvsExpireMinutes()),
      );
      const paymentUrl: string | null = cvsInfo.PaymentURL ?? null;

      if (!orderNumber || !paymentNoVal) {
        this.log.warn({ decryptedJson }, 'ECPay response missing orderNumber or paymentNo');
        return err(DomainError.unexpectedFailure('綠界回傳資料不完整'));
      }

      return ok({
        orderNumber,
        paymentNo: paymentNoVal,
        expireDate: expireDateStr,
        expireAt,
        paymentUrl,
      });
    } catch (e: any) {
      if (e.name === 'TimeoutError' || e.name === 'AbortError') {
        this.log.warn('ECPay request timeout');
        return err(DomainError.unexpectedFailure('綠界連線逾時，請稍後再試'));
      }
      this.log.error({ error: e }, 'Failed to generate ECPay CVS payment code');
      return err(DomainError.unexpectedFailure('建立法幣訂單失敗，請稍後再試'));
    }
  }

  private buildRequestDataPayload(
    merchantId: string,
    merchantTradeNo: string,
    totalAmountTwd: number,
    itemName: string,
    tradeDesc: string,
    returnUrl: string,
    cvsExpireMinutes: number,
  ): string {
    const now = new Date();
    const tradeDate = `${now.getFullYear()}/${pad2(now.getMonth() + 1)}/${pad2(now.getDate())} ${pad2(now.getHours())}:${pad2(now.getMinutes())}:${pad2(now.getSeconds())}`;

    const data = {
      MerchantID: merchantId,
      ChoosePayment: 'CVS',
      OrderInfo: {
        MerchantTradeDate: tradeDate,
        MerchantTradeNo: merchantTradeNo,
        TotalAmount: totalAmountTwd,
        ReturnURL: returnUrl,
        TradeDesc: tradeDesc,
        ItemName: itemName,
      },
      CVSInfo: {
        ExpireDate: cvsExpireMinutes,
        CVSCode: 'CVS',
      },
    };

    return JSON.stringify(data);
  }

  // Atomic counter for same-millisecond sequence numbers
  // NOTE: In-memory static counter is safe under current single-process deployment.
  // If the application needs to scale horizontally in the future, this must be
  // replaced with a Redis atomic INCR or equivalent distributed sequence generator
  // to guarantee uniqueness across processes.
  private static sequenceCounter = 0;
  private static lastTimestampMs = 0;

  /**
   * Generates a unique MerchantTradeNo in format `FD{yyMMddHHmmssSSS}{3-digit-seq}`.
   *
   * Thread-safety in Node.js: this method is fully synchronous (no await points),
   * so it runs atomically within a single event-loop tick. Two concurrent async
   * callers cannot interleave inside this method. Safe for single-process.
   *
   * Multi-process: static counters are process-local — if horizontal scaling is
   * needed, replace with Redis INCR or crypto.randomUUID()-based generation.
   */
  private generateMerchantTradeNo(): string {
    const now = new Date();
    const yy = pad2(now.getFullYear() % 100);
    const MM = pad2(now.getMonth() + 1);
    const dd = pad2(now.getDate());
    const HH = pad2(now.getHours());
    const mm = pad2(now.getMinutes());
    const ss = pad2(now.getSeconds());
    const ms = pad3(now.getMilliseconds());
    const ts = `${yy}${MM}${dd}${HH}${mm}${ss}${ms}`;

    // Synchronized sequence: reset on timestamp change, increment on same ms
    const timestampMs = now.getTime();
    if (timestampMs !== EcpayCvsPaymentService.lastTimestampMs) {
      EcpayCvsPaymentService.lastTimestampMs = timestampMs;
      EcpayCvsPaymentService.sequenceCounter = 0;
    }
    const seq = pad3(EcpayCvsPaymentService.sequenceCounter);
    EcpayCvsPaymentService.sequenceCounter++;

    return `FD${ts}${seq}`;
  }

  private clampCvsExpireMinutes(input: number): number {
    if (input < 1) return 1;
    return Math.min(input, 43200);
  }

  private resolveExpireAt(
    expireDate: string | null,
    fallbackBase: Date,
    expireMinutes: number,
  ): Date {
    const parsed = this.parseExpireAt(expireDate);
    if (parsed) return parsed;
    return new Date(fallbackBase.getTime() + expireMinutes * 60 * 1000);
  }

  private parseExpireAt(expireDate: string | null): Date | null {
    if (!expireDate || expireDate.trim().length === 0) return null;
    try {
      // Format: yyyy/MM/dd HH:mm:ss (Asia/Taipei)
      const parts = expireDate.trim().match(/^(\d{4})\/(\d{2})\/(\d{2}) (\d{2}):(\d{2}):(\d{2})$/);
      if (!parts) return null;
      const taipeiDate = new Date(
        `${parts[1]}-${parts[2]}-${parts[3]}T${parts[4]}:${parts[5]}:${parts[6]}+08:00`,
      );
      return taipeiDate;
    } catch {
      return null;
    }
  }

  private buildTransCodeFailureMessage(transMsg: string): string {
    if (transMsg && transMsg.toLowerCase().includes('decrypt fail')) {
      return `綠界取號失敗：${transMsg}。請確認 ECPAY_STAGE_MODE 是否和 MerchantID/HashKey/HashIV 對應同一環境，並檢查金鑰是否有貼錯或多餘空白。`;
    }
    return `綠界取號失敗：${transMsg}`;
  }
}
